apps/sms/views/conversation/js/compose.js

```diff
@@ -646,73 +619,54 @@ var Compose = (function() {
-        Threads.active.participants :
-        ConversationView.recipients && ConversationView.recipients.numbers;
-      var hasEmailRecipient = recipients ?
-        recipients.some(Utils.isEmailAddress) :
-        false;
-
-      /* Note: in the future, we'll maybe want to force 'mms' from the UI */
-      var newType =
-        hasAttachment() || hasSubject() || hasEmailRecipient || isTextTooLong ?
-        'mms' : 'sms';
-
+      var newType = this.getType();
       if (newType !== state.type) {
         state.type = newType;
         this.emit('type');
       }
     },
 
+    getType() {
+      var typeForcedByLock = state.lock.getType && state.lock.getType();
+      if (typeForcedByLock) {
+        return typeForcedByLock;
+      }
+
+      var isTextTooLong =
+        state.segmentInfo.segments > Settings.maxConcatenatedMessages;
+
+      return isTextTooLong || hasSubject() || hasAttachment() ? 'mms' : 'sms';
+    },
+
     updateEmptyState: function() {
       state.empty = state.emptyMessage && !hasSubject();
     },
 
-    // Send button management
-    /* The send button should be enabled only in the situations where:
-     * - The subject is showing and is not empty (it has text)
-     * - The message is not empty (it has text or attachment)
-    */
     updateSendButton: function() {
-      // should disable if we have no message input
-      var disableSendMessage = state.empty || state.resizing;
-      var messageNotLong = compose.size <= Settings.mmsSizeLimitation;
-
-      /* Bug 1040144: replace ConversationView direct invocation by a
-       * instantiation-time property */
-      var recipients = ConversationView.recipients;
-      var recipientsValue = recipients && recipients.inputValue;
-      var hasRecipients = false;
-
-      // Set hasRecipients to true based on the following conditions:
-      //
-      //  1. There is a valid recipients object
-      //  2. One of the following is true:
-      //      - The recipients object contains at least 1 valid recipient
-      //        - OR -
-      //      - There is >=1 character typed and the value is a finite number
-      //
-      if (recipients &&
-          (recipients.numbers.length ||
-            (recipientsValue && isFinite(recipientsValue)))) {
-
-        hasRecipients = true;
-      }
+      dom.sendButton.disabled = !this.canSend();
+    },
 
-      // should disable if the message is too long
-      disableSendMessage = disableSendMessage || !messageNotLong;
+    updateAttachButton: function() {
+      dom.attachButton.disabled = !this.canEdit();
+    },
+
+    canSend() {
+      var isSendAllowedByLock = !state.lock.canSend || state.lock.canSend();
+
+      // We can't send if message is empty, we're in the process of resizing or
+      // message is already too big.
+      return isSendAllowedByLock && !(this.isEmpty() || this.isResizing ||
+        Settings.mmsSizeLimitation && this.size > Settings.mmsSizeLimitation);
```

Comment:
I think it would be better if you can have local boolean variables with useful names, instead of this big condition.